### PR TITLE
Fixes for PLUGIN-1533 PLUGIN-1469

### PR DESCRIPTION
**Pull request by sgarg-CS**
[PLUGIN-1533](https://cdap.atlassian.net/browse/PLUGIN-1533)

[PLUGIN-1469 ](https://cdap.atlassian.net/browse/PLUGIN-1469)

[PLUGIN-1533]: https://cdap.atlassian.net/browse/PLUGIN-1533?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/Salesforce-batchsink.md [hunk 1]
```diff
@@ -81,6 +81,13 @@ Update - updates existing records based on Id field.
 **Upsert External ID Field:** External id field name. It is used only if operation is upsert.
 The field specified can be either 'Id' or any customly created field, which has external id attribute set.
 
+**Concurrency Mode:** The concurrency mode for the job. The valid values are:<br>
```
Comment (greeshmaswaminathan) [inline docs/Salesforce-batchsink.md]: nit: "The concurrency mode for the bulk job"

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 2]
```diff
@@ -88,6 +90,17 @@ public class SalesforceSinkConfig extends ReferencePluginConfig {
   @Macro
   private String externalIdField;
 
+  @Name(PROPERTY_CONCURRENCY_MODE)
+  @Description("The concurrency mode for the job. The valid values are: \n" +
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: bulk job.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 3]
```diff
@@ -170,6 +184,19 @@ public String getExternalIdField() {
     return externalIdField;
   }
 
+  public String getConcurrencyMode() {
+    return concurrencyMode;
+  }
+
+  public ConcurrencyMode getConcurrencyModeEnum() {
+    try {
+      return ConcurrencyMode.valueOf(concurrencyMode);
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: See if we need null check here and if upgrade from previous version works fine.

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 4]
```diff
@@ -176,7 +177,7 @@ public void createSalesforceListener() throws Exception {
             }
           } else {
             throw new RuntimeException(String.format("Error in meta connect, errorMessage: %s Advice: %s", error,
-                                                     advice));
+                    advice));
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: incorrect formatting

File: docs/Salesforce-batchsink.md (hunk 1 above)
Comment (sgarg-CS) [reply docs/Salesforce-batchsink.md]: done 293aea61e74789e15658e6ba39309865ab94b355

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 2 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: done 293aea61e74789e15658e6ba39309865ab94b355

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Returning the default value: ConcurrencyMode.Parallel if

- the value is null or
- this enum type has no constant with the specified name

Tested with pipeline upgrade and it returns 200 OK. No issues were observed.

293aea61e74789e15658e6ba39309865ab94b355

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java (hunk 4 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Formatted. 293aea61e74789e15658e6ba39309865ab94b355

**Comment by sgarg-CS** [comment]
Also, added a new method to renew the session if the bulk connection resets.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Log the exception to indicate error in mode.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Added error logger in the catch block. 23502c1717880e2be601337d0ce5ca58a32c3fc7

**Comment by sgarg-CS** [comment]
All commits have been squashed.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Add it as an INFO message in the log, and also mention that Parallel is being used as default.

Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Add to the log message that default concurrency mode is used instead. The intention here is to inform that we encountered an issue and took an action.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Updated the INFO log message to 
`LOG.info("Value received was {}. Default concurrency mode i.e. Parallel is used instead.", concurrencyMode);`

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:196]: Squashed all commits.

**Review by greeshmaswaminathan — APPROVED** [review]

**Pull request merged**